doc/designs/quic-design/server/quic-polling.md

```diff
@@ -0,0 +1,1658 @@
+- Our QUIC implementation asks the custom poll method if it understands how to
+- poll those poll descriptors. If not, blocking cannot be supported.
+- When it needs to block, our QUIC implementation uses the custom poll method
+  in either immediate or retained mode based on the poll descriptors reported
+  by the network BIOs provided (internal polling)
+
+### External Polling — Immediate Mode
+
+- Application gets a poll method (default or custom)
+- Application invokes poll() on the poll method on some number of QLSOs, QCSOs, QSSOs
+  and OS sockets, etc.
+- The poll method performs translation to a set of OS resources.
+- The poll method asks the OS to poll/block.
+- The poll method examines the results reported from the OS and performs reverse
+  translation.
+- The poll method poll() call reports the results and returns.
+
+Note that custom poller methods configured on a SSL object are used for internal
+polling (blocking API calls) only. Thus they have no effect on the above
+scenario.
+
+### External Polling — Retained Mode
+
+- Application gets a poll method (default or custom)
+- Application uses the poll method to create a poll group
+- Application registers some number of QLSOs, QCSOs, QSSOs and OS sockets, etc.
+  in the poll group.
+- The poll group caches translations to a set of OS resources. It may create
+  an OS device for fast polling (e.g. epoll) and register these resources
+  with that method.
+- Application polls using the poll group.
+- The poll group asks the OS to poll/block.
+- The poll group examines the results reported from the OS and performs reverse
+  translation.
+- The poll method reports the results and returns.
+
+### External Polling — Immediate Mode Without Event Handling
+
+- Application gets a poll method (default or custom)
+- Application invokes poll() on the poll method on some number of QLSOs, QCSOs,
+  and QSSOs with `NO_HANDLE_EVENTS` set.
+- If the poll method is the default poll method, it knows how to examine
+  QUIC SSL objects for readiness and does so.
+- If the poll method is a custom poll method, it could choose to subdelegate
+  this work to the default poll method, or implement it itself.
+
+Change Notification Callback Mechanism
+--------------------------------------
+
+We propose to allow applications and libssl code to register callbacks for
+lifecycle events on SSL objects, as discussed above. This can be used both by us
+and by applications (e.g. to implement custom poller methods). The advantage
+here is that an SSL object registered into a poll group can be automatically
+unregistered from that poll group when it is freed.
+
+The proposed API is as follows:
+
+```c
+/*
+ * The SSL object is about to be freed (the refcount has reached zero).
+ * The SSL object is still completely healthy until this call returns.
+ * If the SSL object is reffed during a callback, the freeing is cancelled.
+ * The callback then has full responsibility for its lifecycle.
```

Comment:
Do we have to support that? What is the use case?

Reply:
It is so that retained mode poll method implementations know when they can free associated state. So an epoll-based implementation can unregister from epoll, etc. It's quite important.

Reply:
Yeah, I meant the cancellation. OK then.